### PR TITLE
fix potential deadlock in the process of shutdown

### LINKED ISSUE #7765: Segmentation fault when shutdown tiflash
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. Deploy tiflash on S3;
2. Run benchmark and shutdown tiflash

### 2. What did you expect to see? (Required)
TiFlash shutdown successfully.

### 3. What did you see instead (Required)

```
[2023/07/07 06:18:42.076 +00:00] [ERROR] [BaseDaemon.cpp:376] [########################################] [source=BaseDaemon] [thread_id=450]
[2023/07/07 06:18:42.076 +00:00] [ERROR] [BaseDaemon.cpp:377] ["(from thread 404) Received signal Aborted(6)."] [source=BaseDaemon] [thread_id=450]
[2023/07/07 06:18:45.376 +00:00] [ERROR] [BaseDaemon.cpp:569] ["\n       0x7474891\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+122112145]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:220\n  0x7f0d1da36420\t<unknown symbol> [libpthread.so.0+82976]\n  0x7f0d1d79300b\t__GI_raise [libc.so.6+274443]\n                \t/build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:48\n  0x7f0d1d772859\t__GI_abort [libc.so.6+141401]\n                \t/build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:79\n       0x7473fd1\tterminate_handler() [tiflash+122109905]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:0\n  0x7f0d21ee4a13\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\n  0x7f0d21ee49a2\tstd::terminate() [libc++abi.so.1+235938]\n  0x7f0d21f5f8ce\tstd::__1::thread::~thread() [libc++.so.1+461006]\n       0x7e8e31c\tDB::BackgroundProcessingPool::~BackgroundProcessingPool() [tiflash+132703004]\n                \tdbms/src/Storages/BackgroundProcessingPool.cpp:145\n       0x76e3af4\tDB::UniversalPageStorageService::~UniversalPageStorageService() [tiflash+124664564]\n                \tdbms/src/Storages/Page/V3/Universal/UniversalPageStorageService.cpp:308\n       0x76e6eb7\tstd::__1::__shared_ptr_pointer<DB::UniversalPageStorageService*, std::__1::shared_ptr<DB::UniversalPageStorageService>::__shared_ptr_default_delete<DB::UniversalPageStorageService, DB::UniversalPageStorageService>, std::__1::allocator<DB::UniversalPageStorageService> >::__on_zero_shared() [tiflash+124677815]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:267\n       0x76e3de1\tstd::__1::__function::__func<DB::UniversalPageStorageService::create(DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::PSDiskDelegator>, DB::PageStorageConfig const&)::$_3, std::__1::allocator<DB::UniversalPageStorageService::create(DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::PSDiskDelegator>, DB::PageStorageConfig const&)::$_3>, bool ()>::destroy() [tiflash+124665313]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:327\n       0x7e8ef8c\tDB::BackgroundProcessingPool::threadFunction(unsigned long) [tiflash+132706188]\n                \tdbms/src/Storages/BackgroundProcessingPool.cpp:278\n       0x7e8f615\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, DB::BackgroundProcessingPool::BackgroundProcessingPool(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)::$_1> >(void*) [tiflash+132707861]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7f0d1da2a609\tstart_thread [libpthread.so.0+34313]\n                \t/build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477"] [source=BaseDaemon] [thread_id=450]
```

### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: close #7765

Problem Summary: 
The deadlock happens in the following steps:
1. Thread A calls the destructor of `BackgroundProcessingPool`, it modifies `shutdown` to true and called `wake_event.notify_all()`, then wait for all threads in the processing pool to terminate;
2. Thread B in the processing pool just finish a task and calls `BackgroundProcessingPool::tryPopTask` to get more task, because Thread A doesn't modify `shutdown` under any protection of lock, so Thread B doesn't see the change of `shutdown` and continue to call `wake_event.wait_for` for any future notification;

In this case, 
Thread A wait for Thread B to terminate, but Thread B wait for Thread A to notify it, so a deadlock occurs;

This scenario is also depicted in https://en.cppreference.com/w/cpp/thread/condition_variable.
image

This case happened when s3 is enabled because uploading checkpoint is a more time consuming task, so it make the problem more easier to occur.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
restart tiflash multiple times when running workload and see whether there is any error any more.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/BackgroundProcessingPool.cpp [hunk 1]
```diff
@@ -133,7 +133,10 @@ BackgroundProcessingPool::~BackgroundProcessingPool()
 {
     try
     {
-        shutdown = true;
+        {
+            std::unique_lock lock(tasks_mutex);
+            shutdown = true;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/BackgroundProcessingPool.cpp:138]: Can you elaborate about why not acquiring a lock on `tasks_mutex` here will cause a deadlock?

Comment (lidezhu) [reply dbms/src/Storages/BackgroundProcessingPool.cpp:138]: Added in the pr description.

**Review by JaySon-Huang — APPROVED** [review]

**Review by hongyunyan — APPROVED** [review]
LGTM

**Pull request merged**